File: verification/src/main/java/org/msr/mnr/verification/parser/PacketParser.java
package org.msr.mnr.verification.utils;

import java.io.FileReader;
import java.io.IOException;
import java.io.Serializable;
import java.net.DatagramPacket;
import java.util.*;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;
 
public class PacketParser implements Serializable {
    private static final long serialVersionUID = 1L;
    private ParseNode root = null;
    
    public PacketParser(String formatFile) {
        JSONParser jsonParser = new JSONParser();
        try (FileReader reader = new FileReader(formatFile)) {
            JSONObject config = (JSONObject) jsonParser.parse(reader);

            // Parse the packet format
            JSONArray fieldFormat = (JSONArray) config.get("fields");
            root = parseFormat(fieldFormat, null);
        } catch (IOException | ParseException e) {
            e.printStackTrace();
        }
    }

    private ParseNode parseFormat(JSONArray fieldFormat, ParseNode last) {
        ParseNode first = null;

        for (Object nodeObj : fieldFormat) {
            JSONObject node = (JSONObject) nodeObj;

            if (node.size() > 1) {
                ConditionalNode cn = new ConditionalNode(last);

                for (Object conditionStr : node.keySet()) {
                    Condition cond = new Condition((String) conditionStr);
                    JSONArray child = (JSONArray) node.get(conditionStr);
                    cn.children.put(cond, parseFormat(child, null));
                }

                last = cn;
            } else if (node.size() == 1) {
                Object entryObj = node.entrySet().iterator().next();
                Map.Entry<?, ?> entry = Map.Entry.class.cast(entryObj);
                last = new Field(entry.getKey(), entry.getValue(), last);
            } else {
                throw new IllegalArgumentException("Empty field object?");
            }

            if (first == null) {
                first = last;
            }
        }

        return first;
    }

    public void parsePacket(DatagramPacket dp, Packet p) {
        parsePacketHelper(dp.getData(), 0, p, root);
    }

    public void parsePacket(String input, Packet p) {
        byte[] data = new byte[input.length()];
        for (int i = 0; i < input.length(); ++i) {
            data[i] = Byte.parseByte(input.substring(i, i+1), 16);
        }
        parsePacketHelper(data, 0, p, root);
    }

    private int parsePacketHelper(byte[] data, int index, Packet p, ParseNode current) {
        int nextIndex = index;
        if (current instanceof Field) {
            // If it's a simple Field, parse it and set nextIndex
            Field f = (Field) current;
            nextIndex = index + f.length;
            p.set(f.name, data, index, nextIndex);
        } else {
            // If it's a conditional, traverse the correct child
            ConditionalNode cn = (ConditionalNode) current;
            ParseNode child = cn.getChildNode(p);

            if (child != null) {
                nextIndex = parsePacketHelper(data, index, p, child);
            }
        }

        if (current.next != null) {
            nextIndex = parsePacketHelper(data, nextIndex, p, current.next);
        }
        return nextIndex;
    }

    // ===========================================================
    // =================   Helper classes  =======================
    // ===========================================================
    
    private abstract class ParseNode implements Serializable {
        private static final long serialVersionUID = 1L;
        public ParseNode next;

        ParseNode(ParseNode last) {
            if (last != null) {
                last.next = this;
            }
        }
    }
    
    private class Field extends ParseNode {
        private static final long serialVersionUID = 1L;
        public String name;
        public int length;
        
        Field(Object name, Object length, ParseNode last) {
            super(last);
            this.name = (String) name;
            this.length = ((Long) length).intValue();
        }
    }

    private class ConditionalNode extends ParseNode {
        private static final long serialVersionUID = 1L;
        LinkedHashMap<Condition, ParseNode> children = new LinkedHashMap<Condition, ParseNode>();

        ConditionalNode(ParseNode last) {
            super(last);
        }

        ParseNode getChildNode(Packet p) {
            for (Condition c : children.keySet()) {
                if (c.field == null) {
                    return children.get(c);
                }
                if (p.get(c.field).equals(c.value)) {
                    return children.get(c);
                }
            }

            return null;
        }
    }

    private class Condition implements Serializable {
        private static final long serialVersionUID = 1L;
        String field;
        int[] value;

        Condition(String condition) {
            if (condition.equalsIgnoreCase("default")) {
                field = null;
                value = null;
            } else {
                String[] conditionTokens = condition.split("=");
                assert (conditionTokens.length == 2);

                field = conditionTokens[0];
                if (conditionTokens[1].startsWith("0x")) {
                    value = ParseIntArray.fromHexString(conditionTokens[1].substring(2));
                } else {
                    value = ParseIntArray.fromLong(Long.parseLong(conditionTokens[1]));
                }
            }
        }
    }
}

File: C++Verifier/src/utils/PacketParser.cpp
#ifndef PACKETPARSER_H
#define PACKETPARSER_H

#include "../rapidjson/document.h"
#include "../rapidjson/filereadstream.h"
#include "../rapidjson/istreamwrapper.h"
#include "../rapidjson/writer.h"
#include "../rapidjson/stringbuffer.h"
#include "ParserHelper.cpp"

class PacketParser
{
private:
    unique_ptr<ParseNode> root;
    std::vector<string> fieldSequeunce;
public:
    int packetLength = 0;
    PacketParser() {
        root = nullptr;
        packetLength = 0;
    }
    PacketParser(string formatFile) {

        ifstream ifs(formatFile);
        rapidjson::IStreamWrapper isw(ifs);
        rapidjson::Document config;
        config.ParseStream(isw);
        rapidjson::Value& fieldFormat = config["fields"];
        // cout << printValue(fieldFormat) << endl;
        assert(fieldFormat.IsArray());
        packetLength = 0;
        root = parseFormat(fieldFormat, nullptr);
        // for (int i = 0; i <  fieldSequeunce.size(); i++) {
        //     cout << fieldSequeunce[i] << endl;
        // }
    }
    unique_ptr<ParseNode> parseFormat(rapidjson::Value& fieldFormat, ParseNode* last) {
        unique_ptr<ParseNode> first = nullptr;
        // cout << "In parseFormat" << endl;
        for (rapidjson::SizeType i = 0; i < fieldFormat.Size(); i++) {
        
            rapidjson::Value& node = fieldFormat[i];
            int count = 0;
            for (rapidjson::Value::ConstMemberIterator itr = node.MemberBegin(); itr != node.MemberEnd(); ++itr) {
                count += 1;
            }
            if (count > 1) {
                unique_ptr<ConditionalNode> cn = unique_ptr<ConditionalNode>(new ConditionalNode(last));
                for (rapidjson::Value::MemberIterator iter = node.MemberBegin(); iter != node.MemberEnd(); ++iter){
                    string conditionStr = iter -> name.GetString();
                    unique_ptr<Condition> cond = unique_ptr<Condition>(new Condition(conditionStr));
                    rapidjson::Value& child = iter -> value;
                    cn -> children[cond.get()] = parseFormat(child, nullptr);
                }
            } else if (count == 1) {
                last = new Field(node.MemberBegin() -> name.GetString(), node.MemberBegin() -> value.GetInt(), last);
                packetLength += node.MemberBegin() -> value.GetInt();
                fieldSequeunce.push_back(node.MemberBegin() -> name.GetString());
            } else {
                throw std::invalid_argument("Empty field object?");
            }
            if (first == nullptr && last == nullptr) {
                first = nullptr;
            } else if (first == nullptr) {
                first = std::unique_ptr<ParseNode>(last);
            }
        }
        return first;
    }

    ~PacketParser() {
    }

    shared_ptr<Packet> parsePacket(string line) {
        if(startsWith(line, "#") || line.length() == 0 || startsWith(line, "time")) {
            return nullptr;
        }
        return shared_ptr<Packet>(new Packet(line));
    }

    shared_ptr<Packet> parsePacket(unsigned char data[]) {
        shared_ptr<Packet> p(new Packet());
        parsePacketHelper(data, 0, p, root.get());
        return p;
    }

    int parsePacketHelper(unsigned char data[], int index, shared_ptr<Packet> p, ParseNode* current) {
        int nextIndex = index;
        if (Field* f = dynamic_cast<Field*>(current)) {
            nextIndex = index + f-> length;
            p -> set(f -> name, convertByteToString(f -> name, data, index, nextIndex));
            
            // cout << f-> name << endl;
        } else {
            // If it's a conditional, traverse the correct child
            ConditionalNode* cn = dynamic_cast<ConditionalNode*>(current);
            ParseNode* child = cn -> getChildNode(p);

            if (child != nullptr) {
                nextIndex = parsePacketHelper(data, index, p, child);
            }
        }

        if (current -> next != nullptr) {
            nextIndex = parsePacketHelper(data, nextIndex, p, (current -> next).get());
        }
        return nextIndex;
    }

    string printDocument(rapidjson::Document &d) {
        using namespace rapidjson;
        StringBuffer buffer;
        Writer<StringBuffer> writer(buffer);
        d.Accept(writer);

        return buffer.GetString();
    }

    string printValue(rapidjson::Value &val) {
        using namespace rapidjson;
        StringBuffer buffer;
        Writer<StringBuffer> writer(buffer);
        val.Accept(writer);

        return buffer.GetString();
    }

    string getPacketString(shared_ptr<Packet> p) {
        if (root == nullptr) {
            return p -> originalInput();
        }
        std::vector<string> outputVec;
        // cout << "\n vector size initial: " << outputVec.size() << endl;
        // cout << "Sequeunce size: " << fieldSequeunce.size() << endl;
        for (string field: fieldSequeunce) {
            outputVec.push_back(p -> get(field));
            // cout << "Field: " << field << ", Value: " << p -> get(field) << endl;
        }
        // cout << "\n vector size after: " << outputVec.size() << endl;
        return joinVectorString(outputVec, ";");
    }
};


#endif